report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,97 @@
+Learning the syntax of Razor and Flask at the same time provided somewhat of a challenge, and porting to a MVC architecture at the same time didn't make this transition easier.
+
+The arguably biggest part was the implementation of Entity Framework (EF) and porting to PostgreSQL.
+For each part of the model, an entity would have to be established in code, and a context and repository layer for accessing and changing that entity would have to be written.
+Then, on top of that, an API layer was necessary to access and mutate this state.
+On the bright side, this completely eliminated plain text SQL queries in the code itself.
+
+If such a large refactor was to be done some other time, then a more incremental approach might be better.
+If the database was first changed to rely on another DBMS, and _then_ changed to work with EF, then the refactor could become more incremental and thus less monolithic.
+The only problem here is that it would equate to doing double the work, and when a refactor has to be done in a small timeframe, then double work doesn't seem practical.
+
+#### Make sure it all works
+
+In order to ensure that the refactor didn't break anything, tests could be useful.
+Only problem is that the previous tests were an API test script along with Python unit tests.
+
+In order to see immediate results, a Github Action was created to automate running the API test.
+That didn't really work, as it didn't catch all edge cases, so instead the project got configured to be tested with the API simpulator.
+
+Implementation of a proper test suite came later down the line, with
+The team was always aware of the fact that there should be more tests, and were even reminded when things broke in production (like when the register page broke completely).
+Though it is a lesson that the team has already learned in previous projects, here the team learned it once again.
+
+### Evolution
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
+
+Early on in the development process, not many automatic steps were set up in the project pipeline.
+This made evolution much harder than it had to be, as tests would have to be manually run all the time to verify builds, and releases would have to be done manually.
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
+This was, at first, because the course demanded it, but once the team started using automatic steps, then the value was apparent.
+The pipeline was expanded repeatedly, with the steps themselves even being expanded, as is the case with `e2e.yml`.
+
+The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
+A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.
```

Comment:
Maybe reword

> A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.

To be more a clear that it was something we learned.

Something like:

`The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
It is therefore now a focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible. As it would remove the need for doing menial tasks`